File: code/c++/PROT.cpp
#include <iostream>
#include <string>
#include <fstream>
#include <map>
#include <algorithm>
#include <vector>

class PROTSolution
{
private:
    // não codificam para nenhum aminoácido
    std::string strand_mRNA, protein;
    std::vector<std::string> stop_codons = {"UAA", "UAG", "UGA"};
    std::vector<std::string>::iterator it;

    // codificam para algum aminoácido
    std::map<std::string, char> rna_codon_table = {
        {"UUU", 'F'}, {"CUU", 'L'}, {"AUU", 'I'}, {"GUU", 'V'}, 
        {"UUC", 'F'}, {"CUC", 'L'}, {"AUC", 'I'}, {"GUC", 'V'}, 
        {"UUA", 'L'}, {"CUA", 'L'}, {"AUA", 'I'}, {"GUA", 'V'}, 
        {"UUG", 'L'}, {"CUG", 'L'}, {"AUG", 'M'}, {"GUG", 'V'}, 
        {"UCU", 'S'}, {"CCU", 'P'}, {"ACU", 'T'}, {"GCU", 'A'}, 
        {"UCC", 'S'}, {"CCC", 'P'}, {"ACC", 'T'}, {"GCC", 'A'}, 
        {"UCA", 'S'}, {"CCA", 'P'}, {"ACA", 'T'}, {"GCA", 'A'}, 
        {"UCG", 'S'}, {"CCG", 'P'}, {"ACG", 'T'}, {"GCG", 'A'}, 
        {"UAU", 'Y'}, {"CAU", 'H'}, {"AAU", 'N'}, {"GAU", 'D'}, 
        {"UAC", 'Y'}, {"CAC", 'H'}, {"AAC", 'N'}, {"GAC", 'D'}, 
        {"CAA", 'Q'}, {"AAA", 'K'}, {"GAA", 'E'}, {"CAG", 'Q'}, 
        {"AAG", 'K'}, {"GAG", 'E'}, {"UGU", 'C'}, {"CGU", 'R'}, 
        {"AGU", 'S'}, {"GGU", 'G'}, {"UGC", 'C'}, {"CGC", 'R'}, 
        {"AGC", 'S'}, {"GGC", 'G'}, {"CGA", 'R'}, {"AGA", 'R'}, 
        {"GGA", 'G'}, {"UGG", 'W'}, {"CGG", 'R'}, {"AGG", 'R'}, 
        {"GGG", 'G'}};
        
public:
    PROTSolution(){
        read_strand_mRNA();
        translate_rna_into_protein();
    };

    void read_strand_mRNA() {
        std::ifstream input_data("C:/Programming/Rosalind/input/rosalind_prot.txt");

        if (input_data.is_open()) {
            // lendo a fita de mRNA
            std::getline(input_data, strand_mRNA, '\n');
            
            input_data.close();
        } else {
            std::cerr << "Não foi possível abrir o arquivo. " << std::endl;
        }
    }

    // traduzir sequência de dna em proteína
    void translate_rna_into_protein() 
    {
        for (unsigned i = 0; i < strand_mRNA.length(); i += 3) 
        {
            // códon da sequência
            std::string codon = strand_mRNA.substr(i, 3);

            // verificando se o códon da sequência não codifica para algum aminoácido
            it = std::find(stop_codons.begin(), stop_codons.end(), codon);

            // códon codifica para algum aminoácido
            if (it == stop_codons.end()) 
            {
                protein += rna_codon_table[codon];
            }
        }
    }

    void get_protein()
    {
        std::ofstream output_data("C:/Programming/Rosalind/output/PROT.txt");

        if (output_data.is_open()) {
            // proteína
            output_data << protein;
            
            output_data.close();
        } else {
            std::cerr << "Não foi possível abrir o arquivo. " << std::endl;
        }
    }
};

int main()
{
    PROTSolution solution;
    solution.get_protein();

    return 0;
}

File: code/python/PROT.py
class PROTSolution:
    def __init__(self):
        self.strand_mRNA = None
        self.protein = ""
        # não codificam para nenhum aminoácido
        self.stop_codons = ["UAA", "UAG", "UGA"]
        self.aminoacids = ['S', 'L', 'C', 'W', 'E', 'D', 'P', 'V',
                           'N', 'M', 'K', 'Y', 'I', 'Q', 'F', 'R', 'T', 'A', 'G', 'H']
        # codificam para algum aminoácido
        self.rna_codons = [['UCU', 'UCC', 'UCA', 'UCG', 'AGU', 'AGC'], ['UUA', 'UUG', 'CUU', 'CUC', 'CUA', 'CUG'], ['UGU', 'UGC'], ['UGG'], ['GAA', 'GAG'], ['GAU', 'GAC'], ['CCU', 'CCC', 'CCA', 'CCG'], ['GUU', 'GUC', 'GUA', 'GUG'], ['AAU', 'AAC'], [
            'AUG'], ['AAA', 'AAG'], ['UAU', 'UAC'], ['AUU', 'AUC', 'AUA'], ['CAA', 'CAG'], ['UUU', 'UUC'], ['CGU', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'], ['ACU', 'ACC', 'ACA', 'ACG'], ['GCU', 'GCC', 'GCA', 'GCG'], ['GGU', 'GGC', 'GGA', 'GGG'], ['CAU', 'CAC']]

        self.read_strand_mRNA()
        self.translate_rna_into_protein()

    # lendo a fita de mRNA
    def read_strand_mRNA(self):
        with open("C:/Programming/Rosalind/input/rosalind_prot.txt", 'r') as input_data:
            self.strand_mRNA = input_data.read()
            input_data.close()

    # traduzir sequência de dna em proteína
    def translate_rna_into_protein(self):
        for i in range(0, len(self.strand_mRNA), 3):
            codon = self.strand_mRNA[i:i+3]

            # códon codifica para algum aminoácido
            if codon not in self.stop_codons:
                for j, codons in enumerate(self.rna_codons):
                    if codon in codons:
                        self.protein += self.aminoacids[j]

    def get_protein(self):
        with open("C:/Programming/Rosalind/output/PROT.txt", 'w') as output_data:
            output_data.write(self.protein)
            output_data.close()


if __name__ == "__main__":
    solution = PROTSolution()
    solution.get_protein()
